src/MoBi.Core/Services/QuantityValueInSimulationChangeTracker.cs

```diff
@@ -0,0 +1,71 @@
+﻿using System;
+using MoBi.Core.Domain.Model;
+using MoBi.Core.Events;
+using OSPSuite.Core.Domain;
+using OSPSuite.Core.Domain.Builder;
+using OSPSuite.Utility.Events;
+
+namespace MoBi.Core.Services
+{
+   public interface IQuantityValueInSimulationChangeTracker
+   {
+      void TrackChanges(IQuantity quantity, IMoBiSimulation simulation, Action changeAction);
+   }
+
+   public class QuantityValueInSimulationChangeTracker : IQuantityValueInSimulationChangeTracker
+   {
+      private readonly IQuantityToParameterValueMapper _quantityToParameterValueMapper;
+      private readonly IEventPublisher _eventPublisher;
+
+      public QuantityValueInSimulationChangeTracker(IQuantityToParameterValueMapper quantityToParameterValueMapper, IEventPublisher eventPublisher)
+      {
+         _quantityToParameterValueMapper = quantityToParameterValueMapper;
+         _eventPublisher = eventPublisher;
+      }
+
+      public void TrackChanges(IQuantity quantity, IMoBiSimulation simulation, Action changeAction)
+      {
+         storeOriginalQuantityValue(quantity, simulation);
+
+         changeAction();
+
+         checkForOriginalValueRestored(quantity, simulation);
+      }
+
+      private void storeOriginalQuantityValue(IQuantity quantity, IMoBiSimulation simulation)
+      {
+         simulation.AddOriginalQuantityValue(_quantityToParameterValueMapper.MapFrom(quantity));
+         _eventPublisher.PublishEvent(new SimulationStatusChangedEvent(simulation));
```

Comment:
this will be fired even if there are no changes . is this right?

Reply:
yes, that's true. We could count the number of PV's before and after if you think it's worth while to cut down